pkg/syncer/shared/namespace.go

```diff
@@ -105,3 +108,28 @@ func PhysicalClusterNamespaceName(l NamespaceLocator) (string, error) {
 	// keep the namespaces short enough.
 	return fmt.Sprintf("kcp-%s", base36hash[:12]), nil
 }
+
+// IndexByNamespaceLocator is a cache.IndexFunc that indexes namespaces by the namespaceLocator annotation.
+func IndexByNamespaceLocator(obj interface{}) ([]string, error) {
+	metaObj, ok := obj.(metav1.Object)
+	if !ok {
+		return nil, fmt.Errorf("obj is supposed to be a metav1.Object, but is %T", obj)
+	}
+
+	if loc, found, err := LocatorFromAnnotations(metaObj.GetAnnotations()); err != nil {
+		return nil, fmt.Errorf("failed to get locator from annotations: %w", err)
+	} else if !found {
+		return nil, nil
+	} else {
+		nsLocByte, err := json.Marshal(loc)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal locator %#v: %w", loc, err)
+		}
+		return []string{NamespaceLocatorIndexKey(nsLocByte)}, nil
+	}
+}
+
+// NamespaceLocatorIndexKey formats the index key for a namespace locator
+func NamespaceLocatorIndexKey(namespaceLocator []byte) string {
```

Comment:
Hm, this is not very useful. I would expect this to take `*NamespaceLocator` or similar to make it useful to callers.